Fetch Podcast Title From Gutenboarding

allilevine opened the pull request:
#### Changes proposed in this Pull Request

* If we have a podcast ID, request the podcast title from an API endpoint that parses the podcast RSS.

#### Testing instructions

* Checkout this branch
* `ENABLE_FEATURES=anchor-fm-dev` if necessary
* Go to http://calypso.localhost:3000/new?anchor_podcast=3e51070 as both a new user and a signed-in user and check that the podcast title is pre-filled as the site title:
<img width="1412" alt="Screen Shot 2020-12-16 at 4 04 59 PM" src="https://user-images.githubusercontent.com/1689238/102406373-771f3b80-3fb8-11eb-835d-7a6c5f844782.png">

* Also verify that the username (if a new one is created) and the site slug/domain (e.g. allisonspodcast.wordpress.com) reflect the podcast title.

allilevine committed:
Fetch podcast title to pre-fill site title

allilevine committed:
Refactor to fetch podcast title from site title step only

allilevine commented:
@Automattic/luna @mreishus It would be great to get your feedback on whether this is the best approach to requesting and using this data in Gutenboarding. 🙏

mreishus commented:
I'm not able to register a new account with this.  I think it's because your example podcast has "Wordpress" in its name, and that gets passed along as a username hint, somehow..  If I comment out the useEffect, I can register a new account.

![2020-12-17_14-48](https://user-images.githubusercontent.com/937354/102542400-5f5abc80-4077-11eb-936e-1ded3eb555cd.png)
![2020-12-17_14-49](https://user-images.githubusercontent.com/937354/102542419-62ee4380-4077-11eb-842f-281ae88bfb44.png)
![2020-12-17_14-49_1](https://user-images.githubusercontent.com/937354/102542425-65509d80-4077-11eb-89cc-0edba5c55cd0.png)

I still don't understand why or how this would be happening, so maybe I'm wrong. :)

allilevine commented:
> I'm not able to register a new account with this. I think it's because your example podcast has "Wordpress" in its name, and that gets passed along as a username hint, somehow.. If I comment out the useEffect, I can register a new account.

Oh wow! I think you're right, because `http://calypso.localhost:3000/new?anchor_podcast=3e51070` ("Marek's Podcast") works. I tried changing my podcast title but I'm still getting that error, maybe the feed is cached. I updated the testing instructions.

The `username_hint` does use the `siteTitle`:
https://github.com/Automattic/wp-calypso/blob/04bda8b569c73191c0c3897447328ea5f8f3ab31/client/landing/gutenboarding/components/signup-form/index.tsx#L81

Update: my podcast title change is showing up in the feed and I'm no longer getting the error.

pablinos commented:
Is there a way to test for an invalid title ahead of them receiving this error? We should probably recover and just use something else if the user/site name we've chosen for them isn't valid.

mreishus commented:
I agree, I don't think having a podcast that mentions "WordPress" in the title should stop you from signing up.  I think the logic is doing something like this.  Note there is some speculation here:

* When we submit a username to the backend, we also submit a username_hint so in case my username is taken, the backend can suggest alternative usernames.
* "WordPress" is not allowed in usernames for impersonation reasons.
* The backend is overly cynical and rejects any username submission with an invalid username_hint, even if the username itself is okay.

I'm not sure what the best way to fix it is, but maybe the backend needs changing, too (in a diff ticket?)

allilevine commented:
> We should probably recover and just use something else if the user/site name we've chosen for them isn't valid.

> I'm not sure what the best way to fix it is, but maybe the backend needs changing, too (in a diff ticket?)

We're going to try to catch this error on the back-end and recover by falling back to a random username. This is an issue for both regular and Anchor-ified Gutenboarding, but it's more impactful for the latter.

Edit: See D54613-code

allilevine committed:
Update to use useAnchorFmParams()

razvanpapadopol commented:
Thanks @allilevine for creating #48547 to track the site creation error (if the user is already logged in). 

I wasn't sure if we have an issue also for the account creation error so I've just created #48566. Feel free to update or close it if there is already tracked somewhere else.

allilevine committed:
Update feed request with podcast-details endpoint

allilevine commented on client/landing/gutenboarding/onboarding-block/acquire-intent/site-title.tsx:
```diff
@@ -73,6 +79,26 @@ const SiteTitle: React.FunctionComponent< Props > = ( { onSubmit, inputRef } ) =
+			let resp: Promise<object>;
+
+			try {
+				resp = await apiFetch( { path: `https://public-api.wordpress.com/wpcom/v2/podcast-details?url=https://anchor.fm/s/${ encodeURIComponent( anchorFmPodcastId ) }/podcast/rss` } );
```
I'm using `apiFetch()` here because I was getting a JSON syntax error when trying to parse the response from `window.fetch()`. But if we stick with `apiFetch()` then I need to update the code to use its methods.

allilevine commented on client/landing/gutenboarding/onboarding-block/acquire-intent/site-title.tsx:
```diff
@@ -73,6 +79,26 @@ const SiteTitle: React.FunctionComponent< Props > = ( { onSubmit, inputRef } ) =
+				return;
+			}
+
+			let resp: Promise<object>;
```
To do: define the returned data shape.

allilevine committed:
Move podcast title request to window.AppBoot

allilevine committed:
Clean up from moving request

allilevine commented on client/landing/gutenboarding/index.tsx:
```diff
@@ -172,3 +178,29 @@ function ensureRandomizedDesignsAreUpToDate() {
+		path: `https://public-api.wordpress.com/wpcom/v2/podcast-details?url=https://anchor.fm/s/${ encodeURIComponent(
+			anchorFmPodcastId
+		) }/podcast/rss`,
+	} ).then( ( response ) => {
```
`response` needs a Type.

allilevine commented on client/landing/gutenboarding/index.tsx:
```diff
@@ -172,3 +178,29 @@ function ensureRandomizedDesignsAreUpToDate() {
+
+	const anchorFmPodcastId = new URLSearchParams( window.location.search )
+		.get( 'anchor_podcast' )
+		?.replace( /[^a-zA-Z0-9]/g, '' );
```
This is replicated code from path.ts because we can't use hooks here. Maybe we can export a sanitizing function.

yansern replied:
Might be possible to also solve this duplication when it's moved over to a `usePodcastTitle()` hook https://github.com/Automattic/wp-calypso/pull/48430#issuecomment-754632085.

allilevine committed:
Remove try...catch since an unresolved Promise should not be a blocker

mreishus commented:
When my browser makes the request to the podcast-details endpoint, there is some warning text above the json.

![2021-01-04_17-18](https://user-images.githubusercontent.com/937354/103589397-e4840500-4eb0-11eb-8fa6-98b514e22c99.png)

yansern commented:
The placement of setting podcast title and relevant logic at `gutenboarding/index.tsx` feels a little odd.

This can be alleviated by moving it to a hook, say a `usePodcastTitle()` hook, together with the addition of a `useSiteTitle()` hook.

The `usePodcastTitle()` hook is responsible for all the logic related to retrieving the podcast title.
The `useSiteTitle()` hook is responsible for deciding where to retrieve site title / setting initial site title.

Perhaps this is something to be addressed in a subsequent PR.

yansern commented on client/landing/gutenboarding/index.tsx:
```diff
@@ -172,3 +176,29 @@ function ensureRandomizedDesignsAreUpToDate() {
+		return;
+	}
+
+	await apiFetch( {
```
Could the `wpcomRequest()` method can be used in replacement of `apiFetch()`? That way the hostname `public-api.wordpress.com` can be omitted.

allilevine replied:
@yansern I'm not finding that `wpcomRequest()` is currently used outside of the `@automattic/data-stores` package, or part of the package API. Let me know if I missed something! That would be a better option if available.

allilevine committed:
Fetch podcast data and set title with hooks

creativecoder commented:
D54402-code has been merged, so you shouldn't need to sandbox the API for testing this going forward. I've updated the testing instructions in the description accordingly.

allilevine commented:
> The placement of setting podcast title and relevant logic at `gutenboarding/index.tsx` feels a little odd.
> 
> This can be alleviated by moving it to a hook, say a `usePodcastTitle()` hook, together with the addition of a `useSiteTitle()` hook.
> 
> The `usePodcastTitle()` hook is responsible for all the logic related to retrieving the podcast title.
> The `useSiteTitle()` hook is responsible for deciding where to retrieve site title / setting initial site title.
> 
> Perhaps this is something to be addressed in a subsequent PR.

Thanks! It felt a little odd to me too, though having the logic in `gutenboarding/index.tsx` made for a smoother experience. I've moved the logic into the two hooks and I'm calling `useSiteTitle()` from `gutenboard.tsx`. There is a slight delay before the title shows up, but I don't think we can call hooks from `window.AppBoot()`.

allilevine committed:
Fix response type error

allilevine committed:
Change type to interface for PodcastDetailsObject

creativecoder commented on client/landing/gutenboarding/hooks/use-podcast-title.ts:
```diff
@@ -0,0 +1,39 @@
+			path: `https://public-api.wordpress.com/wpcom/v2/podcast-details?url=https://anchor.fm/s/${ encodeURIComponent(
+				anchorFmPodcastId
+			) }/podcast/rss&_fields=title`,
+		} ) as Promise< PodcastDetailsObject > )
```
Rather than use an [overload](https://www.typescriptlang.org/docs/handbook/functions.html#overloads) (`as Promise< PodcastDetailsObject >`) which I think forces the type and overrides the type declaration, you can call the function with the [generic](https://www.typescriptlang.org/docs/handbook/generics.html#working-with-generic-type-variables) specified

```
apiFetch< PodcastDetailsObject >( ...
```

allilevine replied:
Thanks! I made the change.

creativecoder commented on client/landing/gutenboarding/hooks/use-site-title.ts:
```diff
@@ -0,0 +1,23 @@
+	const podcastTitle = usePodcastTitle();
+
+	useEffect( () => {
+		if ( podcastTitle && podcastTitle.length > 1 ) {
```
I think just `if ( podcastTitle )` will do, since an empty string would evaluate to false.

allilevine replied:
I'd added that check because of this selector and note:
https://github.com/Automattic/wp-calypso/blob/8f80f415404c104b1e81d7fc07c177df0548d1ad/client/landing/gutenboarding/stores/onboard/selectors.ts#L30
But it looks like it isn't necessary here, since we don't show the "next" button if that selector doesn't return true.

creativecoder commented on client/landing/gutenboarding/hooks/use-podcast-title.ts:
```diff
@@ -0,0 +1,39 @@
+export default function usePodcastTitle(): string | null {
+	const [ podcastTitle, setPodcastTitle ] = useState< string | null >( '' );
+	const { anchorFmPodcastId } = useAnchorFmParams();
+	interface PodcastDetailsObject {
```
I think `PodcastDetails` should do, as the object is implied by the interface.

creativecoder reviewed (approved):
I left a few minor comments, but this tests well for me and I think using the custom hooks clarify the functionality and purpose of this code.

One UX note is that the podcast title is cut short for some screen sizes. This seems like something we should address (separately), otherwise it can be hard to preview the value that's filled in automatically, because a lot of it is hidden.

<img width="1004" alt="Screen Shot 2021-01-07 at 14 56 24" src="https://user-images.githubusercontent.com/1699996/103945732-3d46de00-50fb-11eb-86a0-b55fb39dd2e3.png">

allilevine committed:
Clean up hooks based on feedback
Rename PodcastDetails, specify a generic instead of an overload, remove
check for podcast title length.
